crates/l2/sequencer/block_producer/payload_builder.rs

```diff
@@ -0,0 +1,290 @@
+    if let Some(gas_used) = gas_limit.checked_sub(context.remaining_gas) {
+        let as_gigas = (gas_used as f64).div(10_f64.powf(9_f64));
+
+        if interval != 0 {
+            let throughput = (as_gigas) / (interval as f64) * 1000_f64;
+            tracing::info!(
+                "[METRIC] BLOCK BUILDING THROUGHPUT: {throughput} Gigagas/s TIME SPENT: {interval} msecs"
+            );
+        }
+    }
+
+    Ok(context.into())
+}
+
+/// Same as `blockchain::fill_transactions` but enforces that the `StateDiff` size  
+/// stays within the blob size limit after processing each transaction.
+pub fn fill_transactions(
+    blockchain: Arc<Blockchain>,
+    context: &mut PayloadBuildContext,
+    store: &Store,
+) -> Result<(), BlockProducerError> {
+    // Two bytes for the len
+    let (mut acc_withdrawals_size, mut acc_deposits_size): (usize, usize) = (2, 2);
+    let mut accounts_info_cache = HashMap::new();
+
+    let chain_config = store.get_chain_config()?;
+    let max_blob_number_per_block: usize = chain_config
+        .get_fork_blob_schedule(context.payload.header.timestamp)
+        .map(|schedule| schedule.max)
+        .unwrap_or_default()
+        .try_into()
+        .unwrap_or_default();
+
+    debug!("Fetching transactions from mempool");
+    // Fetch mempool transactions
+    let (mut plain_txs, mut blob_txs) = blockchain.fetch_mempool_transactions(context)?;
+    // Execute and add transactions to payload (if suitable)
+    loop {
+        // Check if we have enough gas to run more transactions
+        if context.remaining_gas < TX_GAS_COST {
+            debug!("No more gas to run transactions");
+            break;
+        };
+        if !blob_txs.is_empty() && context.blobs_bundle.blobs.len() >= max_blob_number_per_block {
+            debug!("No more blob gas to run blob transactions");
+            blob_txs.clear();
+        }
+        // Fetch the next transactions
+        let (head_tx, is_blob) = match (plain_txs.peek(), blob_txs.peek()) {
+            (None, None) => break,
+            (None, Some(tx)) => (tx, true),
+            (Some(tx), None) => (tx, false),
+            (Some(a), Some(b)) if b < a => (b, true),
+            (Some(tx), _) => (tx, false),
+        };
+
+        let txs = if is_blob {
+            &mut blob_txs
+        } else {
+            &mut plain_txs
+        };
+
+        // Check if we have enough gas to run the transaction
+        if context.remaining_gas < head_tx.tx.gas_limit() {
+            debug!(
+                "Skipping transaction: {}, no gas left",
+                head_tx.tx.compute_hash()
+            );
+            // We don't have enough gas left for the transaction, so we skip all txs from this account
+            txs.pop();
+            continue;
+        }
+
+        // TODO: maybe fetch hash too when filtering mempool so we don't have to compute it here (we can do this in the same refactor as adding timestamp)
+        let tx_hash = head_tx.tx.compute_hash();
```

Comment:
You can use `head_tx.tx.sender()` and that will avoid the hash recomputation (the method uses a cached value for mempool transactions)